### PR TITLE
Adding refresh button and caching browsers.

### PR DESCRIPTION
**Pull request by kthejoker**

### CODE REVIEW

File: BrowserSelect/Program.cs [hunk 1]
```diff
@@ -20,6 +20,8 @@ static class Program
         [STAThread]
         static void Main(string[] args)
         {
+
+            System.Diagnostics.Debug.WriteLine("Test");
```
Comment (zumoshi) [inline BrowserSelect/Program.cs]: ?

File: BrowserSelect/frm_settings.Designer.cs [hunk 2]
```diff
@@ -203,53 +219,70 @@ partial class frm_settings {
...
             this.btn_check_update.Name = "btn_check_update";
-            this.btn_check_update.Size = new System.Drawing.Size(75, 23);
+            this.btn_check_update.Size = new System.Drawing.Size(112, 35);
             this.btn_check_update.TabIndex = 0;
             this.btn_check_update.Text = "check now";
             this.btn_check_update.UseVisualStyleBackColor = true;
             this.btn_check_update.Click += new System.EventHandler(this.btn_check_update_Click);
             // 
-            // chk_check_update
+            // button2
```
Comment (zumoshi) [inline BrowserSelect/frm_settings.Designer.cs]: maybe rename button2 to btn_refresh?

**Comment by kthejoker** [comment]
Whoops, thought I was doing a PR to my own master ...

Anyway, this caches the user's browser list in the settings and only refreshes when the user goes into settings and hits the "Refresh" button (also in PR) next to the browser list.

This reduces startup time of the app by ~80% (10 run avg, 3 Chrome profiles, 2 FF profiles, IE and Edge: 4.06 seconds vs 0.88 seconds).

Possible improvement: staling out the cache over some period of time to occasionally automatically check for browsers/profiles.

File: BrowserSelect/Program.cs (hunk 1 above)
Comment (kthejoker) [reply BrowserSelect/Program.cs]: Sorry, I was statically printing out the JSON being generated to debug the Icon->String->PNG converter, you can remove that line.

File: BrowserSelect/frm_settings.Designer.cs (hunk 2 above)
Comment (kthejoker) [reply BrowserSelect/frm_settings.Designer.cs]: Sorry, kind of new to pull request management - do I make these changes in the original branch and send a new PR? Can I update the existing PR?

Comment (zumoshi) [reply BrowserSelect/frm_settings.Designer.cs]: https://stackoverflow.com/a/16748236/1461004

**Comment by zumoshi** [comment]
I tested it and it is even faster than I thought. The change from the first run (without cache) to second was very noticeable. we can close #40 by merging this.

I only have one complaint, If you had free time to work more on it, If not it's fine. and that is lack of any tooltip/help text. I think users may be confused with how it works. especially since after clicking refresh and closing settings, it doesn't update the list of browsers.

Maybe add a message box after the cache has been invalidated that says `List of browsers updated. new browsers will show up on the next start` or if you had a lot of free time maybe make it refresh the main screen when settings dialogue is closed so changes would apply without needing to restart the app?

**Comment by kthejoker** [comment]
Hmm - absolutely, I see your point. Yes, I will work on this and have it refresh the main form as well.

**Comment by kthejoker** [comment]
Okay, I implemented so that when a user hits refresh the main form updates with the new browser list. Some TODO improvements:

* Currently clears controls and re-adds all instead of being more surgical;
* Also clears and re-adds all browsers even if nothing has changed;

And of course more generally we should have a BindingSource between our browser list and the controls so we can just remove items from the list and have the controls update automatically. Some other day ...

**Comment by zumoshi** [comment]
Neat.

I don't think making a binding is worth it given that this would only be used when a new browser is installed/removed. I would've done it the same way. But I would maybe use a [SuspendLayout](https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.control.suspendlayout?redirectedfrom=MSDN&view=netframework-4.8#System_Windows_Forms_Control_SuspendLayout) call to prevent the in-between state:

image
(this is shown for maybe 2-3 secs with new browsers showing one by one)

**Comment by zumoshi** [comment]
I just noticed a side effect you caused with your last update, `this.Controls.Clear();` removes `btn_help` but doesn't re-add it.
image

But don't worry about it I'll fix it. Thanks again for your work, I'll try to put a release out tonight after fixing #52 .

**Pull request merged**